rust/kernel/error.rs

```diff
@@ -104,3 +105,44 @@ impl From<AllocError> for Error {
         Error::ENOMEM
     }
 }
+
+pub fn from_kernel_result_helper<T>(r: Result<T>) -> T
+where
+    T: TryFrom<c_types::c_int>,
+    T::Error: core::fmt::Debug,
+{
+    match r {
+        Ok(v) => v,
+        Err(e) => T::try_from(e.to_kernel_errno()).unwrap(),
```

Comment:
Just out of curiosity: could this `unwrap` panic?

Reply:
I don't think so, no.

Reply:
Thinking about this a bit more. We can't put `NOPANIC` here, because I suspect that it **can** panic! Consider an `extern C` function that returns `c_char` for example. A contrived example, but hey.

```rust
unsafe extern "C" fn my_callback() -> c_types::c_char {
    from_kernel_result! {
        some_function()?;
        Ok(0)
    }
}
```
Now imagine that `some_function` returns `Err(-EPROBE_DEFER)` which corresponds to `-517`. Since `T` is now a `c_char`, and `-517` can't fit, the function will panic and the kernel dies a horrible death.

Suggestion: we know that `errno`s will never exceed `MAX_ERRNO=4095`. So our negative error codes will allays fit into an `i16` (we can perhaps add a static assertion to make sure this holds always). And then we can use a `Try` trait:

```rust
pub fn from_kernel_result_helper<T>(r: Result<T>) -> T
where
    T: From<i16>,
{
    match r {
        Ok(v) => v,
        Err(e) => T::from(e.to_kernel_errno() as i16),
    }
}
```
Now things will work when returning an integer type that fits (`c_int`, `c_ssize_t`, etc) but `c_char` will fail, which is what we want !
```
error[E0277]: the trait bound `i8: From<i16>` is not satisfied
   --> rust/kernel/error.rs:144:50
    |
110 |   pub fn from_kernel_result_helper<T>(r: Result<T>) -> T
    |          ------------------------- required by a bound in this
111 |   where
112 |       T: From<i16>,
    |          --------- required by this bound in `from_kernel_result_helper`
...
144 |           $crate::error::from_kernel_result_helper((|| {
    |  __________________________________________________^
145 | |             $($tt)*
146 | |         })())
    | |____________^ the trait `From<i16>` is not implemented for `i8`
```

Reply:
SGTM, maybe add a `static_assert!` that `MAX_ERRNO` fits in an `i16`